### PR TITLE
Add SwaggerProducesResponseTypeAttribute and operation filter

### PR DESCRIPTION
**Pull request by tillig**
This is a fix for #173 - putting a response type attribute in place that allows a description to be added to the response type for the purposes of Swagger docs.

I chose to derive the `SwaggerProducesResponseTypeAttribute` from the base ASP.NET core `ProducesResponseTypeAttribute` so you could avoid redundant attributes. Which is to say...

```csharp
// This is redundant, so you don't have to do it.
[ProducesResponseType(typeof(void), 204)]
[SwaggerProducesResponseType(204, description: "No content is returned.")]
public IActionResult MyAction()
{
  // ...
}

// This is all you need - it will work with both standard ASP.NET API description
// as well as Swagger generation.
[SwaggerProducesResponseType(204, description: "No content is returned.")]
public IActionResult OtherAction()
{
  // ...
}
```

I did consider making it an operation filter attribute, but given it led to redundant markup I opted for this way. If that's going to be troublesome or cause confusion, it wouldn't be hard to switch it to an operation filter attribute instead.

### CODE REVIEW

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿using System;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Swashbuckle.SwaggerGen.Annotations
+{
```
Comment (domaindrivendev) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs]: Purely a style preference, but I'm really not a big fan of XML comments, or comments in general for that matter. My philosophy is - strive for expressive code that reveals it's intention on it's own, reserving comments for hacks or non-intuitive workarounds that can't be expressed well with code alone. My bad for not having some contributor guidelines, but any chance you could remove the XML comments :)

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    /// <para>
+    /// This attribute derives from <see cref="ProducesResponseTypeAttribute"/> so you don't
+    /// have to put duplicate/redundant attributes on a controller or action to serve both Swagger
+    /// and standard ASP.NET API description purposes - just use the extended
+    /// Swagger response attribute and it works for both.
+    /// </para>
+    /// </remarks>
+    /// <seealso cref="Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute" />
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true, Inherited = true)]
+    public class SwaggerProducesResponseTypeAttribute : ProducesResponseTypeAttribute
```
Comment (domaindrivendev) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs]: I like the idea of extending ProducesResponseType, but would prefer to still call the attribute _SwaggerResponseAttribute_ as it reads a little better and is more consistent with the existing _SwaggerOperationAttribute_

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApiExplorer;
+using Swashbuckle.Swagger.Model;
+using Swashbuckle.SwaggerGen.Generator;
+
+namespace Swashbuckle.SwaggerGen.Annotations
+{
```
Comment (domaindrivendev) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs]: Remove XML comments as per above, it's hard to read the code with all the comments in the way :)

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        private static void ApplyAttribute(Operation operation, OperationFilterContext context, SwaggerProducesResponseTypeAttribute attribute)
+        {
+            var key = attribute.StatusCode.ToString();
+            Response response;
+            if (!operation.Responses.TryGetValue(key, out response))
+            {
+                response = new Response();
+            }
+
+            response.Description = attribute.Description;
```
Comment (domaindrivendev) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs]: I'm wondering (don't have access to .NET dev environment right now to verify) if the response schema will already be there by virtue of the attribute extending _ProducesResponseType_. Prior to executing filters, the generator will have added responses based on the presence of ApiDescription.SupportedResponseTypes. It depends on the implementation of ApiExplorer, but it's possible that applying the _SwaggerProducesResponseTypeAttribute_ attributes, which extend _ProducesResponseType_, will cause SupportedResponseTypes to be populated. Wondering if you considered this?

**Comment by domaindrivendev** [comment]
A few small comments - but other than that looks awesome!

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs (hunk 1 above)
Comment (tillig) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs]: Before I remove them... XML comments are where IntelliSense sources its doc. While it adds to the code, consumers aren't reading the code at this level.

Are you sure you want them gone?

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs (hunk 4 above)
Comment (tillig) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeFilter.cs]: In unit tests for the filter the whole integration pipeline hasn't run so you get a NullReferenceException without the check. Given I didn't want the filter to assume core behavior lest something inadvertently break, I figured the null check was cheap and reasonable.

File: src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs (hunk 1 above)
Comment (domaindrivendev) [inline src/Swashbuckle.SwaggerGen/Annotations/SwaggerProducesResponseTypeAttribute.cs]: Yep I'd still like to remove them - thanks.

**Comment by tillig** [comment]
Updated per review comments.

**Pull request merged**